xmppserver/src/main/java/org/jivesoftware/openfire/muc/MUCRoom.java

```diff
@@ -2629,47 +2628,105 @@ public void nicknameChanged(MUCOccupant occupant, Presence newPresence, String o
+     * Updates this instance with a subject, without processing it as a subject-change event.
+     *
+     * This method is expected to be used only when a pre-existing room is loaded from a persisted form (such as the
+     * database, or possibly through Serialization in a cluster).
+     *
+     * @param subjectStanza A stanza representing the most recent subject change.
+     */
+    public void initializeSubject(@Nullable final Message subjectStanza)
+    {
+        this.subject = subjectStanza;
+        if (this.subject != null) {
+            this.subject.setTo((JID) null);
+        }
+    }
+
+    /**
+     * Returns the last subject that some occupant set to the room.
+     *
+     * @return the last subject that some occupant set to the room.
+     */
+    public Message getSubjectStanza()
+    {
+        // OF-2163: Prevent modifying the original subject stanza (that can be retrieved by others later) by making a defensive copy.
+        //          This prevents the stanza kept in memory to have the 'to' address for the last user that it was sent to.
+        return this.subject == null ? null : this.subject.createCopy();
+    }
+
     /**
      * Returns the last subject that some occupant set to the room.
      *
      * @return the last subject that some occupant set to the room.
      */
     public String getSubject() {
-        return subject;
+        final Message subjectStanza = getSubjectStanza();
+        return subjectStanza == null ? null : subjectStanza.getSubject();
     }
 
     /**
-     * Sets the last subject that some occupant set to the room. This message will only be used
-     * when loading a room from the database.
+     * Sets the current subject of the room.
      *
-     * @param subject the last known subject of the room (cannot be {@code null}).
+     * @param subject the new subject of the room.
+     * @deprecated Use {@link #changeSubject(Message, MUCOccupant)} or {@link #initializeSubject(Message)} instead.
      */
+    @Deprecated(forRemoval = true) // Remove in or after Openfire 5.2.0
     public void setSubject(String subject) {
-        this.subject = subject;
+        try {
+            final Message roomSubject = new Message();
+            roomSubject.setType(Message.Type.groupchat);
+            roomSubject.setID(UUID.randomUUID().toString());
+            roomSubject.setFrom(this.getJID());
+
+            // Add the subject to the 'subject' element of the message. Ensure that this element is always present (even
+            // when empty), as MUC joins require 'subject' element to be present.
+            if (subject != null && !subject.isEmpty()) {
+                roomSubject.setSubject(subject);
+            } else {
+                roomSubject.getElement().addElement("subject");
+            }
+
+            changeSubject(roomSubject, this.selfOccupantData);
+        } catch (ForbiddenException e) {
+            Log.warn("Unable to change the subject of room {}", this.getJID(), e);
```

Comment:
Why is this the only exception that could happen? Surely there could (should?) be database and other things that could occur and be trapped and bubbled up?